### PR TITLE
Revise LinkControl suggestions UI to use MenuItem

### PR DESCRIPTION
**Pull request by richtabor**
## What?
Part of https://github.com/WordPress/gutenberg/issues/50885, as part of a greater [LinkControl refresh](https://github.com/WordPress/gutenberg/issues/50891) effort. Note that this should follow https://github.com/WordPress/gutenberg/pull/50977. 

- [x]  Try using `<MenuItem>` for items instead of custom buttons + styles. 
- [x]  Clean up search results metrics to match the mockups proposed in https://github.com/WordPress/gutenberg/issues/50891
- [x]  Refactor unused CSS
- [x]  Try adding back permalink info per [this discussion](https://github.com/WordPress/gutenberg/issues/50891#issuecomment-1561440083) 
- [x]  Tests

## Testing Instructions

1. Open a Post or Page.
2. Insert a Navigation Block.
3. Add a link item. 
4. See changes. 
5. Insert a Paragraph block.
6. Create an inline link. 
7. See changes. 

## Screenshots or screencast 

https://github.com/WordPress/gutenberg/assets/1813435/3ceb1600-3c64-4342-ab2f-054e641c9257

Co-authored-by: Dave Smith <444434+getdave@users.noreply.github.com>

### CODE REVIEW

File: packages/block-editor/src/components/link-control/constants.js
Comment (getdave) [inline packages/block-editor/src/components/link-control/constants.js:11]: Is this a presentational change?

**Comment by richtabor** [comment]
> Try adding back permalink info per https://github.com/WordPress/gutenberg/issues/50891#issuecomment-1561440083

We can use the MenuItem info prop to pass the permalink.

File: packages/base-styles/_variables.scss
Comment (getdave) [inline packages/base-styles/_variables.scss]: @richtabor are you sure you want to make this change in _this_ PR? If you anticipate it potentially being a large UI change then it could be better to pull this out into it's own PR in order that it could be reverted relatively easily if required.

**Comment by getdave** [comment]
@richtabor I made some changes here and fixed the tests. Let me know what you think of it.

We need to choose and implement the visual "states" for the suggestions list:

- hover
- focus
- selected (use keyboard arrow keys)

I've hacked something in but if you have a clear idea on these that would help.

**Comment by richtabor** [comment]
> We need to choose and implement the visual "states" for the suggestions list:

Wouldn't the MenuItem component styles be fine here?

**Comment by getdave** [comment]
Uh oh! What happened? Looks like we need to rebase `trunk` to get this back to normal.

**Comment by jasmussen** [comment]
Thanks for working on this. From the video shared, this looks great, though I'm seeing disabled save/cancel buttons in trunk:

Screenshot 2023-06-21 at 08 22 10

Was there an initial intent to omit those until an actionable state?

I'm also not seeing the down arrow key to select a menu item work, but perhaps that's related to the focus style comment? Thanks for the work, this is important!

**Review by getdave — COMMENTED** [review]
We need to address:

- ~keyboard arrow keys do not work~ add _visual_ states for aria-selected
- ~pressing `TAB` when in input results in selection of that result~ [this is present in `trunk`](https://github.com/WordPress/gutenberg/issues/51732).

File: packages/block-editor/src/components/link-control/style.scss
Comment (getdave) [inline packages/block-editor/src/components/link-control/style.scss]: Removing this completely removed all "selected" styling from the list. I restored it in https://github.com/WordPress/gutenberg/pull/50978/commits/cecd7c8ba190c768f1717287cb8b6f04c0885f54

File: packages/block-editor/src/components/link-control/style.scss
Comment (getdave) [inline packages/block-editor/src/components/link-control/style.scss:103]: I think the usage of a negative margin here necessitates a comment to explain _why_ we're doing it. @richtabor can you recall why you added this?

**Comment by getdave** [comment]
> though I'm seeing disabled save/cancel buttons in trunk:

I think that will be handled in a seperate PR.

@jasmussen @richtabor are you happy with this visual selected state? As you can see when the item is _not_ selected all the padding looks fine, but when it _is_ it looks like a mistake (at least to me) as the selected item looks very close to the edge of the popover.

Screen Shot 2023-06-21 at 11 28 51

File: packages/base-styles/_variables.scss
Comment (getdave) [reply packages/base-styles/_variables.scss]: Let's pull this out into it's own PR as a follow up.

Comment (getdave) [reply packages/base-styles/_variables.scss]: https://github.com/WordPress/gutenberg/pull/51733

**Review by getdave — COMMENTED** [review]
I'm generally happy with this code-wise except from:

- the padding issue I flagged
- the need to pull out https://github.com/WordPress/gutenberg/pull/51733

I can't approve this as I've co-authored it so I'll wait for a ✅ from someone else.

File: packages/block-editor/src/components/link-control/constants.js
Comment (richtabor) [reply packages/block-editor/src/components/link-control/constants.js:11]: Yes, it changes the type from "URL" to "link" visually.

File: packages/block-editor/src/components/link-control/style.scss
Comment (richtabor) [reply packages/block-editor/src/components/link-control/style.scss:103]: I added a comment. It's needed because the link control field needs margin all around it to render properly without search results.

**Comment by richtabor** [comment]
> are you happy with this visual selected state? 

Any reason why we wouldn't use the existing focus state for selection? 

CleanShot 2023-06-21 at 11 22 37

**Comment by getdave** [comment]
> > are you happy with this visual selected state?
> 
> Any reason why we wouldn't use the existing focus state for selection?
> 
> CleanShot 2023-06-21 at 11 22 37

I'm fine with whatever you think is right. But what I will say is that I guess "focus" !== "selection" in this kind of UI because the actual programmatic focus appears to remain on the `<input>`. In this case we should be careful about mixing up the two types of state.

I think that's intentional. If it's not it's a bug 😟

**Comment by richtabor** [comment]
> the actual programmatic focus appears to remain on the <input>. In this case we should be careful about mixing up the two types of state.

Good point. I've updated the selected state to resemble the latest from the Inserter and Wayfinder styles. 

CleanShot 2023-06-21 at 12 00 27

**Comment by richtabor** [comment]
> Was there an initial intent to omit those until an actionable state

Yes, there's nothing to save while a link has not been added (it's disabled in trunk until a link has been added, or changes are made). We'd only have the Cancel button, which we don't include on other popovers anyhow.

**Comment by jameskoster** [comment]
> But what I will say is that I guess "focus" !== "selection" in this kind of UI because the actual programmatic focus appears to remain on the <input>

Just to note this is a widespread issue that needs a holistic solution. See: https://github.com/WordPress/gutenberg/issues/51582

> I've updated the selected state to resemble the latest from the Inserter and Wayfinder styles.

Seems like a fair interim.

File: packages/base-styles/_variables.scss
Comment (getdave) [inline packages/base-styles/_variables.scss]: Needs rebase to remove this from the changeset.

Comment (richtabor) [reply packages/base-styles/_variables.scss]: Rebased https://github.com/WordPress/gutenberg/pull/50978/commits/ba7681b718840128eb329579a5556c7c83ffdc8b

**Review by getdave — APPROVED** [review]
I'm generally happy with these changes now. 

We just need to determine whether the e2e test failures are legit or not before we can merge this. Remember as Link UI impacts a large part of the codebase it can sometimes trigger unexpected test failures.

**Comment by getdave** [comment]
@tellthemachines Let's bring this into 6.3. It was missing that label as you identified in https://github.com/WordPress/gutenberg/pull/52517#issuecomment-1631854793

Thank you 🙏

**Comment by tellthemachines** [comment]
Done! This should really have been cherry-picked before #51712, would have avoided a ton of conflict-solving.

**Comment by getdave** [comment]
> Done! This should really have been cherry-picked before #51712, would have avoided a ton of conflict-solving.

Absolutely. Apologies that this happened. Just an oversight I'm afraid. We had a lot of concurrent PRs and this one slipped through the net. Appreciate your time and effort resolving 🙇‍♂️

**Pull request merged**
